File: benchmarks/c/06_prime_sieve.c
// Prime Sieve Benchmark - C Implementation
// Sieve of Eratosthenes algorithm

#include <stdio.h>
#include <stdbool.h>

#define N 10000

int main() {
    // Initialize sieve (true = potentially prime)
    bool sieve[N + 1];
    for (int i = 0; i <= N; i++) {
        sieve[i] = true;
    }
    sieve[0] = false;
    sieve[1] = false;
    
    // Sieve of Eratosthenes
    for (int p = 2; p * p <= N; p++) {
        if (sieve[p]) {
            // Mark all multiples of p as not prime
            for (int i = p * p; i <= N; i += p) {
                sieve[i] = false;
            }
        }
    }
    
    // Count primes
    int count = 0;
    for (int i = 0; i <= N; i++) {
        if (sieve[i]) {
            count++;
        }
    }
    
    printf("primes up to %d: %d\n", N, count);
    return 0;
}

File: benchmarks/python/06_prime_sieve.py
def main():
    n = 10000
    sieve = [1] * (n + 1)
    sieve[0] = 0
    sieve[1] = 0

    p = 2
    while p * p <= n:
        if sieve[p] == 1:
            i = p * p
            while i <= n:
                sieve[i] = 0
                i += p
        p += 1

    count = 0
    i = 0
    while i <= n:
        if sieve[i] == 1:
            count += 1
        i += 1

    print(count)


if __name__ == "__main__":
    main()
